### ISSUE TITLE
Must escape environment variables containing % and other special INI characters

### ISSUE THREAD

**Issue by mmerickel** [Pylons/plaster_pastedeploy#9]

**Comment by bertjwregeer** [comment]
Closed by #10 and d9711edb37432e36f64d142edd474f7fee8b98cf

**Comment by tigerfoot** [comment]
Even with 0.4.1 release I still get those stange messages : 

' ' '
manuel serve
Starting monitor for PID 10069.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.6/site-packages/hupper/ipc.py", line 322, in spawn_main
    func(**kwargs)
  File "/usr/lib/python3.6/site-packages/hupper/worker.py", line 237, in worker_main
    func(*spec_args, **spec_kwargs)
  File "/usr/lib/python3.6/site-packages/pyramid/scripts/pserve.py", line 32, in main
    return command.run()
  File "/usr/lib/python3.6/site-packages/pyramid/scripts/pserve.py", line 229, in run
    app = loader.get_wsgi_app(app_name, config_vars)
  File "/usr/lib/python3.6/site-packages/plaster_pastedeploy/__init__.py", line 131, in get_wsgi_app
    global_conf=defaults)
  File "/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py", line 247, in loadapp
    return loadobj(APP, uri, name=name, **kw)
  File "/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py", line 271, in loadobj
    global_conf=global_conf)
  File "/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py", line 296, in loadcontext
    global_conf=global_conf)
  File "/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py", line 320, in _loadconfig
    return loader.get_context(object_type, name, global_conf)
  File "/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py", line 454, in get_context
    section)
  File "/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py", line 476, in _context_from_use
    object_type, name=use, global_conf=global_conf)
  File "/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py", line 406, in get_context
    global_conf=global_conf)
  File "/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py", line 296, in loadcontext
    global_conf=global_conf)
  File "/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py", line 320, in _loadconfig
    return loader.get_context(object_type, name, global_conf)
  File "/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py", line 413, in get_context
    defaults = self.parser.defaults()
  File "/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py", line 68, in defaults
    defaults[key] = self.get('DEFAULT', key) or val
  File "/usr/lib64/python3.6/configparser.py", line 800, in get
    d)
  File "/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py", line 95, in before_get
    value, defaults)
  File "/usr/lib64/python3.6/configparser.py", line 394, in before_get
    self._interpolate_some(parser, option, L, value, section, defaults, 1)
  File "/usr/lib64/python3.6/configparser.py", line 444, in _interpolate_some
    "found: %r" % (rest,))
configparser.InterpolationSyntaxError: Error in file /ioda/home/bruno/git/geo/geo-api3/production.ini: '%' must be followed by '%' or '(', found: '%s %s'
Server with PID 10069 exited with code 1.
' ' ' 

Seems related to content in development/production.ini

'''
[formatters]
keys = generic
[formatter_generic]
format = "%(asctime)s %(levelname)-5.5s [%(name)s][%(threadName)s] %(message)s"
'''

I would like to get ride of those troubles, but don't know where to start ;-)

**Comment by mmerickel** [comment]
Sorry @tigerfoot but there's just not enough info here for me to diagnose what's going on. Do you have a reproducible example?

The issue is not with the logger sections in your example. If it were we'd be seeing many more errors reported and I'm using it myself with such things in my ini file with no issues.

**Comment by tigerfoot** [comment]
@mmerickel yeap sorry, your comment make sense, I will wednesday try to open an issue with most of usefull information, with a reproductible way, as much I can (that the trouble of being n00bs in some technology).

**Comment by hanula** [comment]
The issue is still present. Happens with inheriting configs and `'%'` in ENV variables.
Easiest to reproduce is to add `monkeypatch.setenv('PLASTER_BAR', '%bar')` to [test_foreign_config](https://github.com/Pylons/plaster_pastedeploy/blob/760c2eb57c284b911b31270bb9d5f70eb6fc2627/tests/test_get_wsgi_app_settings.py#L28).

**Comment by mmerickel** [comment]
I just dug into this and you're right, it's still an issue with the apis that are relying on pastedeploy to parse the file. The issue is due to a bug in pastedeploy that was fixed long ago but it still unreleased. I do not see a simple workaround for this.

I think for now I'm just going to need to filter out any env vars with a bare `%` in them instead of trying to fix them.

**Comment by bowlofeggs** [comment]
I see a test failing with this error on Python 3.6.3:

```
$ py.test-3
================================================ test session starts =================================================
platform linux -- Python 3.6.3, pytest-3.2.3, py-1.5.2, pluggy-0.4.0                                                  
rootdir: /home/rbarlow/rpmbuild/BUILD/plaster_pastedeploy-0.4.2, inifile: setup.cfg                                   
plugins: cov-2.5.1                                         
collected 73 items                                                                                                    
                                                                                                                      
tests/test_configdict.py ..                                
tests/test_get_settings.py ....................            
tests/test_get_wsgi_app.py .....................           
tests/test_get_wsgi_app_settings.py ..F...                 
tests/test_get_wsgi_filter.py .........                    
tests/test_get_wsgi_server.py .........                    
tests/test_loader.py .                                     
tests/test_setup_logging.py .....                                                                                     
                                                                                                                      
====================================================== FAILURES ====================================================== 
__________________________________________ TestFullURI.test_foreign_config ___________________________________________

self = <tests.test_get_wsgi_app_settings.TestFullURI object at 0x7fde08d8e5c0>                                         
                                                                                                                      
    def test_foreign_config(self):                         
>       result = self.loader.get_wsgi_app_settings('test_foreign_config')                                              
                                                                                                                      
test_get_wsgi_app_settings.py:29:                                                                                     
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  
../src/plaster_pastedeploy/__init__.py:201: in get_wsgi_app_settings                                                  
    global_conf=defaults)    
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:261: in appconfig                                           
    global_conf=global_conf) 
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:296: in loadcontext                                         
    global_conf=global_conf)                                                                                          
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:320: in _loadconfig                                         
    return loader.get_context(object_type, name, global_conf)                                                          
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:454: in get_context                                         
    section)                                               
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:476: in _context_from_use                                   
    object_type, name=use, global_conf=global_conf)        
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:406: in get_context
    global_conf=global_conf)
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:296: in loadcontext
    global_conf=global_conf)
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:320: in _loadconfig
    return loader.get_context(object_type, name, global_conf)
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:413: in get_context
    defaults = self.parser.defaults()
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:68: in defaults
    defaults[key] = self.get('DEFAULT', key) or val
/usr/lib64/python3.6/configparser.py:800: in get
    d)
/usr/lib/python3.6/site-packages/paste/deploy/loadwsgi.py:95: in before_get
    value, defaults)
/usr/lib64/python3.6/configparser.py:394: in before_get
    self._interpolate_some(parser, option, L, value, section, defaults, 1)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <configparser.BasicInterpolation object at 0x7fde09b2c908>
parser = <paste.deploy.loadwsgi.NicerConfigParser object at 0x7fde08dbd0f0>, option = 'ENV_TMUX_PANE', accum = []
rest = '%4', section = 'DEFAULT'
map = ChainMap({}, {}, OrderedDict([('here', '/home/rbarlow/rpmbuild/BUILD/plaster_pastedeploy-0.4.2/tests/sample_configs'),...), ('basepath', '/home/rbarlow/rpmbuild/BUILD/plaster_pastedeploy-0.4.2/tests/sample_configs'), ('glob', 'overr
ide')]))
depth = 1

    def _interpolate_some(self, parser, option, accum, rest, section, map,
                          depth):
        rawval = parser.get(section, option, raw=True, fallback=rest)
        if depth > MAX_INTERPOLATION_DEPTH:
            raise InterpolationDepthError(option, section, rawval)
        while rest:
            p = rest.find("%")
            if p < 0:
                accum.append(rest)
                return
            if p > 0:
                accum.append(rest[:p])
                rest = rest[p:]
            # p is no longer used
            c = rest[1:2]
            if c == "%":
                accum.append("%")
                rest = rest[2:]
            elif c == "(":
                m = self._KEYCRE.match(rest)
                if m is None:
                    raise InterpolationSyntaxError(option, section,
                        "bad interpolation variable reference %r" % rest)
                var = parser.optionxform(m.group(1))
                rest = rest[m.end():]
                try:
                    v = map[var]
                except KeyError:
                    raise InterpolationMissingOptionError(
                        option, section, rawval, var) from None
                if "%" in v:
                    self._interpolate_some(parser, option, accum, v,
                                           section, map, depth + 1)
                else:
                    accum.append(v)
            else:
                raise InterpolationSyntaxError(
                    option, section,
                    "'%%' must be followed by '%%' or '(', "
>                   "found: %r" % (rest,))
E               configparser.InterpolationSyntaxError: Error in file /home/rbarlow/rpmbuild/BUILD/plaster_pastedeploy-0.4.2/tests/sample_configs/test_config_included.ini: '%' must be followed by '%' or '(', found: '%4'

/usr/lib64/python3.6/configparser.py:444: InterpolationSyntaxError
======================================== 1 failed, 72 passed in 2.38 seconds =========================================
```

**Comment by hanula** [comment]
Same as in my case, it's `TMUX_PANE` env variable from `tmux` which is set to `{num}%`. Really annoying. 
`set -u TMUX_PANE` should help until the issue is fixed.

**Comment by mmerickel** [comment]
Yeah sorry folks I know this is a problem and it's one I can't fix in plaster_pastedeploy unless I just disable the env-var feature entirely (which maybe I should do). The bug is one that we've exposed in pastedeploy. Basically we escape the strings, then tell pastedeploy to use these escaped strings, but then it ends up unescaping them and causing them to get interpolated.

In an effort to fix this, a couple weeks ago I actually took over maintainership of pastedeploy itself and plan to release a fix to it which would enable plaster_pastedeploy's current code to actually work but I have had no time to actually do this as taking over a long-standing project that comes with a bunch of cruft is a large undertaking. It's in mercurial in bitbucket right now on top of all other maintainer issues.

I will say I'm tempted to just drop the env-var feature from plaster_pastedeploy as perhaps it was a bad idea to add it.
